File: src/prod/src/ServiceModel/ReplicatorQueueStatus.cpp
// ------------------------------------------------------------
// Copyright (c) Microsoft Corporation.  All rights reserved.
// Licensed under the MIT License (MIT). See License.txt in the repo root for license information.
// ------------------------------------------------------------

#include "stdafx.h"

using namespace std;
using namespace Common;
using namespace ServiceModel;

ReplicatorQueueStatus::ReplicatorQueueStatus()
    : queueUtilizationPercentage_(0)
    , queueMemorySize_(0)
    , firstSequenceNumber_(-1)
    , completedSequenceNumber_(-1)
    , committedSequenceNumber_(-1)
    , lastSequenceNumber_(-1)
{
}

ReplicatorQueueStatus::ReplicatorQueueStatus(
    uint queueUtilizationPercentage,
    LONGLONG queueMemorySize,
    FABRIC_SEQUENCE_NUMBER firstSequenceNumber,
    FABRIC_SEQUENCE_NUMBER completedSequenceNumber,
    FABRIC_SEQUENCE_NUMBER committedSequenceNumber,
    FABRIC_SEQUENCE_NUMBER lastSequenceNumber)
    : queueUtilizationPercentage_(queueUtilizationPercentage)
    , queueMemorySize_(queueMemorySize)
    , firstSequenceNumber_(firstSequenceNumber)
    , completedSequenceNumber_(completedSequenceNumber)
    , committedSequenceNumber_(committedSequenceNumber)
    , lastSequenceNumber_(lastSequenceNumber)
{
}

ReplicatorQueueStatus::ReplicatorQueueStatus(ReplicatorQueueStatus && other)
    : queueUtilizationPercentage_(other.queueUtilizationPercentage_)
    , queueMemorySize_(other.queueMemorySize_)
    , firstSequenceNumber_(other.firstSequenceNumber_)
    , committedSequenceNumber_(other.committedSequenceNumber_)
    , completedSequenceNumber_(other.completedSequenceNumber_)
    , lastSequenceNumber_(other.lastSequenceNumber_)
{
}

ReplicatorQueueStatus const & ReplicatorQueueStatus::operator = (ReplicatorQueueStatus && other)
{
    if (this != &other)
    {
        this->queueUtilizationPercentage_ = other.queueUtilizationPercentage_;
        this->queueMemorySize_ = other.queueMemorySize_;
        this->firstSequenceNumber_ = other.firstSequenceNumber_;
        this->committedSequenceNumber_ = other.committedSequenceNumber_;
        this->completedSequenceNumber_ = other.completedSequenceNumber_;
        this->lastSequenceNumber_ = other.lastSequenceNumber_;
    }

    return *this;
}

ErrorCode ReplicatorQueueStatus::ToPublicApi(
    __in ScopedHeap & heap, 
    __out FABRIC_REPLICATOR_QUEUE_STATUS & publicResult) const
{
    UNREFERENCED_PARAMETER(heap);
    publicResult.QueueUtilizationPercentage = this->queueUtilizationPercentage_;
    publicResult.CommittedSequenceNumber = this->committedSequenceNumber_;
    publicResult.CompletedSequenceNumber = this->completedSequenceNumber_;
    publicResult.FirstSequenceNumber = this->firstSequenceNumber_;
    publicResult.LastSequenceNumber = this->lastSequenceNumber_;
    publicResult.QueueMemorySize = this->queueMemorySize_;
    
    return ErrorCodeValue::Success;
}

ErrorCode ReplicatorQueueStatus::FromPublicApi(
    __in FABRIC_REPLICATOR_QUEUE_STATUS & publicResult)
{
    this->committedSequenceNumber_ = publicResult.CommittedSequenceNumber;
    this->completedSequenceNumber_ = publicResult.CompletedSequenceNumber;
    this->firstSequenceNumber_ = publicResult.FirstSequenceNumber;
    this->lastSequenceNumber_ = publicResult.LastSequenceNumber;
    this->queueMemorySize_ = publicResult.QueueMemorySize;
    this->queueUtilizationPercentage_ = publicResult.QueueUtilizationPercentage;

    return ErrorCode();
}

void ReplicatorQueueStatus::WriteTo(TextWriter& w, FormatOptions const &) const
{
    w.Write(ToString());
}

wstring ReplicatorQueueStatus::ToString() const
{
    return wformatString(
        "queueUtilizationPercentage = '{0}', FirstSequenceNumber = '{1}', committedSequenceNumber = '{2}', completedSequenceNumber = '{3}', lastSequenceNumber = '{4}', QueueMemorySize = '{5}'",
       queueUtilizationPercentage_, firstSequenceNumber_, committedSequenceNumber_, completedSequenceNumber_, lastSequenceNumber_, queueMemorySize_);
}

File: src/prod/src/managed/Api/src/System/Fabric/query/ReplicatorQueueStatus.cs
// ------------------------------------------------------------
// Copyright (c) Microsoft Corporation.  All rights reserved.
// Licensed under the MIT License (MIT). See License.txt in the repo root for license information.
// ------------------------------------------------------------

namespace System.Fabric.Query
{
    using System.Fabric.Interop;

    /// <summary>
    /// <para>Provides various statistics of the queue used in the Service Fabric Replicator.</para>
    /// </summary>
    /// <remarks>
    /// <para>Depending on the role of the replicator (<see cref="System.Fabric.ReplicaRole.Primary" /> 
    /// or <see cref="System.Fabric.ReplicaRole.ActiveSecondary" />), the properties in this type mean different things.</para>
    /// </remarks>
    public sealed class ReplicatorQueueStatus
    {
        /// <summary>
        /// <para>Initializes a new instance of the <see cref="System.Fabric.Query.ReplicatorQueueStatus" /> class.</para>
        /// </summary>
        public ReplicatorQueueStatus()
        {
        }

        /// <summary>
        /// <para>Gets the utilization of the queue.</para>
        /// </summary>
        /// <value>
        /// <para>The utilization of the queue.</para>
        /// </value>
        /// <remarks>
        /// <para>A value of ‘0’ indicates that the queue is empty and a value of ‘100’ indicates that the queue is full.</para>
        /// </remarks>
        public long QueueUtilizationPercentage { get; private set; }

        /// <summary>
        /// <para>Gets the number of virtual memory bytes being consumed by the queue.</para>
        /// </summary>
        /// <value>
        /// <para>The number of virtual memory bytes being consumed by the queue.</para>
        /// </value>
        public long QueueMemorySize { get; private set; }

        /// <summary>
        /// <para>See remarks.</para>
        /// </summary>
        /// <value>
        /// <para>The first sequence number.</para>
        /// </value>
        /// <remarks>
        ///   <list type="number">
        ///     <item>
        ///       <description>
        ///         <para>
        ///           <see cref="System.Fabric.ReplicaRole.Primary" /> – Represents the smallest sequence number of the operation that is 
        /// present in the queue. It’s value will be the same as <see cref="System.Fabric.Query.ReplicatorQueueStatus.CompletedSequenceNumber" />, 
        /// since the primary replicator discards operations once it receives an acknowledgement from all the secondary replicas.</para>
        ///       </description>
        ///     </item>
        ///     <item>
        ///       <description>
        ///         <para>
        ///           <see cref="System.Fabric.ReplicaRole.ActiveSecondary" /> – Represents the sequence number of the first operation 
        /// that is available in the queue.</para>
        ///       </description>
        ///     </item>
        ///   </list>
        /// </remarks>
        public long FirstSequenceNumber { get; private set; }

        /// <summary>
        /// <para>See remarks.</para>
        /// </summary>
        /// <value>
        /// <para>The completed sequence number.</para>
        /// </value>
        /// <remarks>
        ///   <list type="number">
        ///     <item>
        ///       <description>
        ///         <para>
        ///           <see cref="System.Fabric.ReplicaRole.Primary" /> – Represents the highest sequence number for which <b>all</b> the 
        /// secondary replicas have applied the operation.</para>
        ///       </description>
        ///     </item>
        ///     <item>
        ///       <description>
        ///         <para>
        ///           <see cref="System.Fabric.ReplicaRole.ActiveSecondary" /> – Represents the highest sequence number that has been 
        /// applied by the user service replica.</para>
        ///       </description>
        ///     </item>
        ///   </list>
        /// </remarks>
        public long CompletedSequenceNumber { get; private set; }

        /// <summary>
        /// <para>See remarks.</para>
        /// </summary>
        /// <value>
        /// <para>The committed sequence number.</para>
        /// </value>
        /// <remarks>
        ///   <list type="number">
        ///     <item>
        ///       <description>
        ///         <para>
        ///           <see cref="System.Fabric.ReplicaRole.Primary" /> – Represents the highest sequence number for which a <b>quorum</b> 
        /// of secondary replicas have applied the operation.</para>
        ///       </description>
        ///     </item>
        ///     <item>
        ///       <description>
        ///         <para>
        ///           <see cref="System.Fabric.ReplicaRole.ActiveSecondary" /> – Represents the highest sequence number that has been 
        /// received by the secondary replicator from the primary.</para>
        ///       </description>
        ///     </item>
        ///   </list>
        /// </remarks>
        public long CommittedSequenceNumber { get; private set; }

        /// <summary>
        /// <para>See remarks.</para>
        /// </summary>
        /// <value>
        /// <para>The last sequence number.</para>
        /// </value>
        /// <remarks>
        ///   <list type="number">
        ///     <item>
        ///       <description>
        ///         <para>
        ///           <see cref="System.Fabric.ReplicaRole.Primary" /> – Represents the latest sequence number of the operation that 
        /// is available in the queue.</para>
        ///       </description>
        ///     </item>
        ///     <item>
        ///       <description>
        ///         <para>
        ///           <see cref="System.Fabric.ReplicaRole.ActiveSecondary" /> – Represents the latest sequence number of the operation 
        /// that is available in the queue.</para>
        ///       </description>
        ///     </item>
        ///   </list>
        /// </remarks>
        public long LastSequenceNumber { get; private set; }

        internal static unsafe ReplicatorQueueStatus CreateFromNative(NativeTypes.FABRIC_REPLICATOR_QUEUE_STATUS* nativeEntryPoint)
        {
            return new ReplicatorQueueStatus
            {
                QueueUtilizationPercentage = nativeEntryPoint->QueueUtilizationPercentage,
                FirstSequenceNumber = nativeEntryPoint->FirstSequenceNumber,
                CompletedSequenceNumber = nativeEntryPoint->CompletedSequenceNumber,
                CommittedSequenceNumber = nativeEntryPoint->CommittedSequenceNumber,
                LastSequenceNumber = nativeEntryPoint->LastSequenceNumber,
                QueueMemorySize = nativeEntryPoint->QueueMemorySize,
            };
        }
    }
}